chart/values.schema.yaml

```diff
@@ -322,6 +322,31 @@ properties:
 
   system_domain: {type: string}
 
+  instance_group_condition:
+    type: object
+    additionalProperties:
+      type: object
+      additionalProperties:
+        oneOf:
+        - { type: boolean }
+        - { type: string }
+
+  limits:
```

Comment:
Please use the `sizing.` key.
e.g. `sizing.diego-cell.limits.memory`.

Reply:
This changes the structure in a way which makes it impossible (I believe) to have default values (per igroup, and global). Also note that your structure limits sizing to the igroup/pod level ... The current structure (ignoring limits vs sizing) easily allows for defaults, and allows limits at the job/container level.

```
limits:
  memory:
    '$default': 10Mi
    api:
      '$default': 100Mi
      cloud_controller_ng: 2Gi
```

whereas with

```
sizing:
  diego-cell:
    limits:
      memory: 4Gi
```

slotting in defaults looks to much harder, if not impossible.

Reply:
> slotting in defaults looks to much harder, if not impossible.

It could work the same way, but I agree that it looks convoluted because we have the additional job level to consider too:

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    limits:
      memory:
        '$default': 128Mi
        cloud_controller_ng: 2Gi
```

I'm especially unhappy about having `$default` and `$defaults` properties, and how they have different internal structure. But the nesting is also getting too deep.

Of course the defaults are mostly in the internal config file as a user override would not bother with defaults normally and just specify e.g.

```
sizing:
  api:
    limits:
      memory:
        cloud_controller_ng: 4Gi
```

I still can't really make up my mind: What is more important, seeing all the sizing/limits information for a specific instance group in one place, or seeing all the memory limits grouped together in one place, and all the instance counts in another?

Conceptually seeing all the information about the pods in one place makes sense. But when I'm looking at the memory requirements of the cluster, having all the limits close together is more helpful than having them scattered between all the other settings.

But I think we need to settle on one ordering mechanism: if we have a separate `limits` structure, then maybe we should split `sizing` into multiple trees too? Why combine instance counts and affinity settings? But then, `limits` should not combine `memory` and `cpu` limits either; otherwise why not fold it into `sizing`?

One other difference is that `sizing` stops at the pod (instance group) level, while `limits` are applied to the container (job).

After writing all this I'm even less able to decide which one feels best...

Reply:
Oh, one more thought: maybe having per-instance group defaults doesn't really make much sense. You just have one global default, and then have to override any job that needs extra memory.

I know I suggested this originally, but now I can't think of an actual use case where setting different per-pod limits is required.

Reply:
I am ok with just having a global default.

Reply:
Using the bosh manifest deployment as an example, more often than not, it's useful to see all the information for one instance group in one place. Things that make you have to change context in the yaml file (like addons or variables) are painful. As an analogy, even though the variables are defined as an array in the deployment manifest, I never care about looking at them at-a-glance. It's always their context that matters.

If you need info grouped by limits, or disk size, or privileges, you should be able to script it easily.

---

For the defaults - can you please explain why per-ig defaults are useful?
The containers in a pod have very different memory needs - I think each individual container should have a value.

Reply:
So, based on that 

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    limits:
      memory:
        cloud_controller_ng: 2Gi
```

or 

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    cloud_controller_ng:
      limits:
        memory: 2Gi
```

I prefer the second form (`group/job/'limit'/'memory'`) over (`group/'limits'/'memory'/job`).

Reply:
> please explain why per-ig defaults are useful

They came from Jan, and he now has no idea either regarding their usefulness.

Reply:
> I prefer the second form (`group/job/'limits'/'memory'`) over (`group/'limits'/'memory'/job`).

That doesn't really work as there are other properties below the group already:

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    instances: 3
    affinity: { ... }
    limits:
      memory:
        cloud_controller_ng: 2Gi
```

So you would still need something like this:

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    instances: 3
    affinity: { ... }
    jobs:
      cloud_controller_ng:
        limits:
          memory: 4Gi
      cloud_controller_clock:
        limits:
          memory: 512Mi
```

(yes, I know `cloud_controller_clock` is in the `scheduler` pod, just an example of setting 2 limits for one group).

So as soon as you set multiple limits I think grouping limits together works better:

```
sizing:
  '$defaults':
    limits:
      memory: 10Mi
  api:
    instances: 3
    affinity: { ... }
    limits:
      memory:
        cloud_controller_ng: 4Gi
        cloud_controller_clock: 512Mi
```

Of course once you add in CPU limits and maybe CPU and memory requests it all becomes a toss-up again:

```
sizing:
  '$defaults':
    limits:
      cpu: 2
      memory: 256Mi
    requests:
      memory: 64Mi
  api:
    instances: 3
    affinity: { ... }
    limits:
      cpu:
        cloud_controller_ng: 4
      memory:
        cloud_controller_ng: 4Gi
        cloud_controller_clock: 512Mi
    requests:
      memory:
        cloud_controller_ng: 2Gi
        cloud_controller_clock: 256Mi
```

I think I still slightly prefer this grouping, as you may need/want not set non-default cpu limits for the same jobs that you want to specify non-default memory limits for.

Reply:
Ah. I forgot/missed that we already have `sizing`, from `values.yaml`. :facepalm:
Ok. Now reading the remainder of this comment.

If @viovanov is ok with this last proposed structure I am ok with it as well.

Reply:
This last one looks ok - but I still don't think the defaults are useful. e.g. 256Mi for the CC is not enough, but too much for log cache, or nats or other such things. If we already have real-world numbers, why bother with the defaults?

```
sizing:
  '$defaults':
    limits:
      cpu: 2
      memory: 256Mi
    requests:
      memory: 64Mi
  api:
    instances: 3
    affinity: { ... }
    limits:
      cpu:
        cloud_controller_ng: 4
      memory:
        cloud_controller_ng: 4Gi
        cloud_controller_clock: 512Mi
    requests:
      memory:
        cloud_controller_ng: 2Gi
        cloud_controller_clock: 256Mi
```

Reply:
Ok, how about leaving out the defaults for now, and if it comes to it us wanting them later on it should be no trouble to add them based on the above.

Reply:
> I still don't think the defaults are useful. e.g. 256Mi for the CC is not enough, but too much for log cache, or nats or other such things.

The numbers here are just placeholders, but remember they are `limits`, not `requests`.

They exist to for the benefit of *other* pods on the same node, in case any of our containers has a memory leak. Just like an insurance policy, we ideally want non of our limits to ever get triggered, we just want to prevent one container from consuming all memory on the node, essentially performing a DoS attack on the other pods.

So to me it totally makes sense to specify a pretty generous limit as the default, maybe even 512Mi, and only specify larger limits for the few containers that absolutely need it (e.g the CC and the Java components like credhub and UAA).

If NATS only uses 128Mi, it doesn't matter at all that we specify a limit of `512Mi`, it will never trigger, so it behaves exactly the same as a limit of `256Mi`. or `160Mi`. It only means we won't kill NATS if it happens to unexpectedly use `320Mi`, but I don't consider that an issue either.

Again these are limits, not requests, so no memory is "reserved" for it; it just is the number at which the OOM killer will enter a container.

So I feel having just a shorter list of the exceptionally large limits is more maintainable than a list that includes every single container. Remember that we will then have to update the list whenever a job is added anywhere in the system.

Finally, I wonder how all this applies to init containers. I think we need a mechanism to tell Quarks to set some kind of limit on them as well.

Reply:
No idea if it applies to init containers at all. I suspect not.

Reply:
> No idea if it applies to init containers at all. I suspect not.

I didn't think so either. This was a comment for @viovanov that we may need additional support for this. Because init containers run pre-start scripts, which could leak memory and never terminate... So as an admin, I may still want to prevent unlimited growth.